### PR TITLE
feat(frontend): explicitly label conversions in Ethereum tokens

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

# Changes

# Tests

### CODE REVIEW

File: src/frontend/src/icp-eth/derived/cketh.derived.ts [hunk 1]
```diff
@@ -117,3 +118,20 @@ export const ckErc20HelperContractAddress: Readable<OptionEthAddress> = derived(
 	([$ckEthMinterInfoStore, $ethereumTokenId]) =>
 		toCkErc20HelperContractAddress($ckEthMinterInfoStore?.[$ethereumTokenId])
 );
+
+/**
+ * The minter address.
+ */
+const ckMinterAddress: Readable<OptionEthAddress> = derived(
```
Comment (peterpeterparker) [inline src/frontend/src/icp-eth/derived/cketh.derived.ts:125]: I'm not convinced we should introduce a new global store for this scoped feature. On one hand, our idea is to deprecate tokenWillFallback; on the other, adding more global stores can impact performance. So unless this is truly used globally and not just in one or two functions, I don't think it should be a global store. WDYT?

**Review by peterpeterparker — COMMENTED** [review]
Just one feedback, did not reviewed the PR yet.

File: src/frontend/src/icp-eth/derived/cketh.derived.ts (hunk 1 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/icp-eth/derived/cketh.derived.ts]: It is not using tokenWillFallback (or a derived), I think. But in any case, I understand what you mean with not adding more stores. Please check last commit, I moved the store logic inside the component.

File: src/frontend/src/eth/components/transactions/Transaction.svelte [hunk 2]
```diff
@@ -23,22 +38,84 @@
 	let timestamp: number | undefined;
 	let displayTimestamp: number | undefined;
 
+	let ckMinterInfoAddresses: (EthAddress | undefined)[];
+	$: $ckEthMinterInfoStore,
+		$ethereumToken,
+		(() => {
+			const ckMinterInfoData = $ckEthMinterInfoStore?.[$ethereumToken.id];
```
Comment (peterpeterparker) [inline src/frontend/src/eth/components/transactions/Transaction.svelte:45]: This feels like it doesn't belong here.

- We do not want to map and iterate over the minter addresses on each transaction render for performance reasons.

- I understand now why you primarily created a global store, but even with such a store, we would still need to map here.

- However, the component should remain as naive as possible, and we are doing the opposite by adding logic here. "It's just rendering".

Spontaneously, I feel like we need something similar to IC side; we need some kind of mapper.

Basically:

sortedTransactions + ckMinterAddress.recude => map in a utils = upgraded list => <Transaction />

Do you know what I mean? What do you think?

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:45]: Ah, you mean to add a way to add the prop `type` directly to the transactions list?

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:45]: - Either the transaction component calls a map or the all combined list goes through a mapper.

- `ckMinterAddress.recude` should be performed only once.

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: Maybe have a look at last commit, is that it?

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: Yeah exactly, I think it's better. That way the component is naive and the heavy load is done has few times as possible without have global performance impact.

Can you add tests for the utils?

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: Yes, makes sense!

I was waiting to create tests, until I confirmed this was the change

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: mmmm you have broke something, a bunch of code you removed previously is back in the component

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: nvm maybe I oversee something. let me have another look

File: src/frontend/src/eth/components/transactions/Transaction.svelte [hunk 3]
```diff
@@ -2,43 +2,84 @@
...
 
-	$: ({ from, value, timestamp, displayTimestamp } = transaction);
+	$: ({ value, timestamp, displayTimestamp, uiType: type } = transaction);
 
-	let type: EthTransactionType;
-	$: type = from?.toLowerCase() === $ethAddress?.toLowerCase() ? 'send' : 'receive';
+	let ckTokenSymbol: string;
+	$: ckTokenSymbol =
+		nonNullish($tokenId) && isSupportedEthTokenId($tokenId)
+			? `ck${$tokenSymbol}`
```
Comment (peterpeterparker) [inline src/frontend/src/eth/components/transactions/Transaction.svelte:40]: Not sure why this `ck${$tokenSymbol}` mapping, can you briefly explain what information we are missing?

File: src/frontend/src/eth/components/transactions/Transaction.svelte [hunk 4]
```diff
@@ -2,43 +2,84 @@
...
+			: ($token as Erc20Token).twinTokenSymbol ?? '';
```
Comment (peterpeterparker) [inline src/frontend/src/eth/components/transactions/Transaction.svelte:41]: If `$tokenId` can potentially be nullish, couldn't `$token` be also potentially nullish?

File: src/frontend/src/eth/components/transactions/Transaction.svelte [hunk 5]
```diff
@@ -2,43 +2,84 @@
...
+					: $i18n.receive.text.receive;
 
 	let icon: ComponentType;
-	$: icon = type === 'send' ? IconSend : IconReceive;
+	$: icon =
+		(type === 'withdraw' || type === 'deposit') && pending
+			? IconConvert
+			: type === 'withdraw'
+				? IconMint
+				: type === 'deposit'
```
Comment (peterpeterparker) [inline src/frontend/src/eth/components/transactions/Transaction.svelte:75]: It's correct, deposit = burn?

File: src/frontend/src/eth/components/transactions/Transaction.svelte [hunk 6]
```diff
@@ -2,43 +2,84 @@
...
+					? IconBurn
+					: type === 'send'
+						? IconSend
+						: IconReceive;
 
 	let amount: BigNumber;
-	$: amount = type == 'send' ? value.mul(BigNumber.from(-1)) : value;
```
Comment (peterpeterparker) [inline src/frontend/src/eth/components/transactions/Transaction.svelte:41]: Oopsie

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 2 above)
Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: Added few minor questions as I had another look. Hope you don't mind. But this is good, thanks a ton!

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 5 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:75]: yes, they represent the same "scope" in the ck-conversions, but I am using the naming of the [documentation](https://github.com/dfinity/ic/blob/master/rs/ethereum/cketh/docs/cketh.adoc) for the transaction types. For the icons, I was planning already to change the name of the icons in another PR, so that we call them in another way, representing this

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 4 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:41]: yes, that `nonNullish` is just because `isSupportedEthTokenId` does not accept `undefined`... what would you suggest?

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 3 above)
Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:40]: I think there is something, like not a link but there is already a piece of code somewhere that does something similar....

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:40]: I don't remember where so not a blocker but, ....

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 4 above)
Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:41]: If the `$token` is potentially `undefined` then the following cast is maybe more accurate:

```
($token as Erc20Token | undefined)?.twinTokenSymbol ?? '';
```

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:41]: But if it cannot be undefined, it's fine.

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 5 above)
Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:75]: Thanks for the explanation.

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 4 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:41]: I will use your suggestion, but I add a TODO about changing it when `$token` will become something "better" eheh

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 3 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:40]: I searched and I cannot find anything, do you have some clue on what to search? Or maybe about which flow?

P.S. the alternative would be to make an array of ckETH data (prod, staging and local) and use `.find` to compare against their `twinToken` property. That would be the completly correct way

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:40]:
```typescript
// PSEUDO CODE

export const CKETH_DATA: IcCkInterface[] = [
	...(nonNullish(CKETH_LOCAL_DATA) ? [CKETH_LOCAL_DATA] : []),
	...(nonNullish(CKETH_STAGING_DATA) ? [CKETH_STAGING_DATA] : []),
	...(nonNullish(CKETH_IC_DATA) ? [CKETH_IC_DATA] : [])
];

CKETH_DATA.find(({ twinToken: { id } }) => id === $tokenId)?.symbol;
```

Comment (peterpeterparker) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:40]: Ah, couldn't we use `twinTokenSymbol` or `Erc20Contract`?

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:40]: for ETH and SepoliaETH we don't have the property `twinTokenSymbol`, we never assign it. Shall I add it (in another PR)?

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 5 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte:77]: Changed icons name with https://github.com/dfinity/oisy-wallet/pull/2347

File: src/frontend/src/eth/components/transactions/Transaction.svelte (hunk 3 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/eth/components/transactions/Transaction.svelte]: Doing with PRs:

- https://github.com/dfinity/oisy-wallet/pull/2348
- https://github.com/dfinity/oisy-wallet/pull/2350
-
